### PR TITLE
haxe.macro.Expr.hx documentation

### PR DESCRIPTION
**Pull request by markknol**
Parts based on the documentation of haxe.macro.Type.hx documentation.
Please double check. 

It has full stops :smile:

### CODE REVIEW

File: std/haxe/macro/Expr.hx [hunk 1]
```diff
@@ -25,21 +25,62 @@ package haxe.macro;
...
 }
 #end
 
+/**
+	Represents a constant.
+	@see http://haxe.org/manual/expression-constants.html
+**/
 enum Constant {
+	/**
+		Represents a integer literal.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: an

File: std/haxe/macro/Expr.hx [hunk 2]
```diff
@@ -25,21 +25,62 @@ package haxe.macro;
...
+	**/
 	CFloat( f : String );
+	
+	/*
+		Represents a string literal.
+	**/
 	CString( s : String );
+	
+	/*
+		Represents a indentifier literal.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: That sounds strange, just say "Represents an identifier".

File: std/haxe/macro/Expr.hx [hunk 3]
```diff
@@ -25,21 +25,62 @@ package haxe.macro;
...
+	**/
 	CIdent( s : String );
+	
+	/*
+		Represents a regular expression literal.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Maybe we should give an example for all of these. This one in particular.

File: std/haxe/macro/Expr.hx [hunk 4]
```diff
@@ -169,154 +213,603 @@ enum Unop {
 	OpNegBits;
 }
 
+/**
+	Represents a AST node for the AST.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: redundant AST

File: std/haxe/macro/Expr.hx [hunk 5]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 
+/**
+	Represents a AST node identical to `Expr`, but it allows constraining the 
+	type of accepted expressions.
+	@see http://haxe.org/manual/macro-ExprOf.html
+**/
 typedef ExprOf<T> = Expr;
 
+/**
+	Represents a case for the AST.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: "A case for the AST" sounds like a blog post title. Let's say "Represents a switch case".

File: std/haxe/macro/Expr.hx [hunk 6]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	@see http://haxe.org/manual/expression-switch.html
+**/
 typedef Case = {
+	/**
+		The value expressions of the case.
+	**/
 	var values : Array<Expr>;
+	/**
+		The guard expressions of the case. Guards restrict cases.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: "Guards restrict cases" is technically correct but sounds kinda strange here. Let's lose it.

File: std/haxe/macro/Expr.hx [hunk 7]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	Represents a variable for the AST.
+	@see http://haxe.org/manual/expression-var.html
+**/
 typedef Var = {
+	/**
+		The name of the variable.
+	**/
 	name : String,
+	/**
+		The type of the variable.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: It's better to say type-hint. Also we should add ", if available" to these Null-fields.

File: std/haxe/macro/Expr.hx [hunk 8]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 	expr : Null<Expr>
 }
 
+/**
+	Represents a catch for the AST.
+	@http://haxe.org/manual/expression-try-catch.html
+**/
 typedef Catch = {
+	/**
+		The name of the catch.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: It's the name of the catch _variable_.

File: std/haxe/macro/Expr.hx [hunk 9]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	type : ComplexType,
+	/**
+		The expression of the catch.
+	**/
 	expr : Expr
 }
 
+/**
+	Represents kind of a node for the AST.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: I don't know if there's an article missing here but "kind of a node" sounds really stupid. :D

File: std/haxe/macro/Expr.hx [hunk 10]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	EArray( e1 : Expr, e2 : Expr );
+
+	/**
+		Binary operator `e1 op e2`.
+	**/
 	EBinop( op : Binop, e1 : Expr, e2 : Expr );
+
+	/**
+		Field access on `e` according to `field`.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: EField is much simpler than TField, it's really just `e.field`.

File: std/haxe/macro/Expr.hx [hunk 11]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+		* ++e (op = OpIncrement, postFix = false)
+		* --e (op = OpDecrement, postFix = false)
+		* -e (op = OpNeg, postFix = false)
+		* !e (op = OpNot, postFix = false)
+		* ~e (op = OpNegBits, postFix = false)
+	**/
 	EUnop( op : Unop, postFix : Bool, e : Expr );
+
+	/**
+		A variable declaration `var v` or `var v = expr`.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: In this AST there can be multiple variables for `var x, y` notation, hence the Array argument.

File: std/haxe/macro/Expr.hx [hunk 12]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	EVars( vars : Array<Var> );
+
+	/**
+		A function declaration.
+	**/
 	EFunction( name : Null<String>, f : Function );
+
+	/**
+		A block declaration `{exprs}`.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: I missed this in Type.hx, this is not really a declaration. "A block of expressions" would be better.

File: std/haxe/macro/Expr.hx (hunk 9 above)
Comment (markknol) [inline std/haxe/macro/Expr.hx]: Ahum https://github.com/HaxeFoundation/haxe/blob/development/std/haxe/macro/Type.hx#L807 :wink:

File: std/haxe/macro/Expr.hx [hunk 13]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	EBreak;
+
+	/**
+		A `continue` expression.
+	**/
 	EContinue;
+
+	/**
+		A `untyped e` expression.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: This should be "An" even though what follows is source code (I think).

File: std/haxe/macro/Expr.hx [hunk 14]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 	ECheckType( e : Expr, t : ComplexType );
+
+	/**
+		A `@m e` expression.
+	**/
 	EMeta( s : MetadataEntry, e : Expr );
 }
 
+/**
+	Represents a complex type for the AST.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: To be honest the whole "complex" terminology is quite silly. It was originally named like that but we really shouldn't mention it in the documentation unless we refer to ComplexType itself. For all intents and purposes a complex type is just a type syntax.

File: std/haxe/macro/Expr.hx (hunk 9 above)
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Definitely a missing "the" in that case. :P

Also "in the AST" sounds much better than "for the AST".

File: std/haxe/macro/Expr.hx [hunk 15]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 	TFunction( args : Array<ComplexType>, ret : ComplexType );
+
+	/**
+		Represents an anonymous structure type.
+		@see http://haxe.org/manual/types-anonymous-structure.html
+	**/
 	TAnonymous( fields : Array<Field> );
+
+	/**
+		Represents the parent type.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: While the thought of types having parents is entertaining, this is actually about parentheses, e.g. in `(Int -> Void) -> Void` or something.

File: std/haxe/macro/Expr.hx [hunk 16]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	TParent( t : ComplexType );
+
+	/**
+		Represents the structure to the type inheritance structure.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Wat?

File: std/haxe/macro/Expr.hx [hunk 17]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	TExtend( p : Array<TypePath>, fields : Array<Field> );
+
+	/**
+		Represents the optional type.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: the -> an

File: std/haxe/macro/Expr.hx [hunk 18]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	var name : String;
+
+	/**
+		Optional parameters of the type path.
+	**/
 	@:optional var params : Array<TypeParam>;
+
+	/**
+
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: sub is set on module sub-type access: pack.Module.Type has name = Module, sub = Type.

File: std/haxe/macro/Expr.hx [hunk 19]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	/**
+		
+	**/
 	@:optional var sub : Null<String>;
 }
 
+/**
+	Represents a type parameter for the AST.
+**/
 enum TypeParam {
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: This one is really wrong. Note that it's an enum, not a structural type.

Haxe allows expressions in concrete type parameters, e.g. `new YourType<["hello", "world"]>`. In that case the value is TPExpr while in the normal case it's TPType.

File: std/haxe/macro/Expr.hx [hunk 20]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	TPExpr( e : Expr );
 }
 
+/**
+	Represents a type parameter declaration for the AST.
+**/
 typedef TypeParamDecl = {
+	/**
+		The name of the type parameter declaration.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Lose the "declaration" here and in the other fields.

File: std/haxe/macro/Expr.hx [hunk 21]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	Represents a function for the AST.
+**/
 typedef Function = {
+	/**
+		A list of function arguments.
+	**/
 	var args : Array<FunctionArg>;
+
+	/**
+		The return type of the function.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: type-hint, if available

File: std/haxe/macro/Expr.hx [hunk 22]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	var name : String;
+
+	/**
+		Whether or not the function argument is optional.
+	**/
 	@:optional var opt : Bool;
+
+	/**
+		The type of the function argument.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: type-hint, if available

File: std/haxe/macro/Expr.hx [hunk 23]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 	@:optional var doc : Null<String>;
+
+	/**
+		The access modifiers of the field. By default fields have private access.
+		@see http://haxe.org/manual/class-field-access-modifier.html
+	**/
 	@:optional var access : Array<Access>;
+
+	/**
+		The type of the field.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: "type" is really overloaded, let's stick to "kind". I know it's a bit confusing...

File: std/haxe/macro/Expr.hx [hunk 24]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+
+	/**
+		Static access modifier. 
+	**/
 	AStatic;
+
+	/**
+		Override access modifier, required when a field is declared which also 
+		exists on a parent class. This modifier is only allowed on 
+		method fields.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Mmh... this is about syntax and I believe we actually allow parsing `override var x`, so this information is somewhat out of place.

File: std/haxe/macro/Expr.hx [hunk 25]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	AMacro;
 }
 
+/**
+	Represents the field type for the AST.
+**/
 enum FieldType {
+	/**
+		Represents a variable structure.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: Why "structure"?

File: std/haxe/macro/Expr.hx [hunk 26]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
 	FFun( f : Function );
+
+	/**
+		Represents a property with getter and setter structure.
+	**/
 	FProp( get : String, set : String, ?t : Null<ComplexType>, ?e : Null<Expr> );
 }
 
+/**
+	Represents a type definition for the AST.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: This one is unrelated to the AST itself.

File: std/haxe/macro/Expr.hx [hunk 27]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	@:optional var params : Array<TypeParamDecl>;
+
+	/**
+		Whether or not the type is extern.
+	**/
 	@:optional var isExtern : Bool;
+
+	/**
+		The type kind of the type definition.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: type kind? :P

File: std/haxe/macro/Expr.hx [hunk 28]
```diff
@@ -169,154 +213,603 @@ enum Unop {
...
+	**/
 	TDStructure;
+
+	/**
+		Represents a class kind.
+	**/
 	TDClass( ?superClass : TypePath, ?interfaces : Array<TypePath>, ?isInterface : Bool );
+
+	/**
+		Represents an alias kind.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: I don't know why this one was named alias. At the very least we should say "alias/typedef".

File: std/haxe/macro/Expr.hx [hunk 29]
```diff
@@ -324,24 +817,65 @@ enum TypeDefKind {
...
+	/**
+		Instantiates an error with given message and position.
+	**/
 	public function new(m,p) {
 		this.message = m;
 		this.pos = p;
 	}
+
+	/**
+		String representation of the error.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: This is a method so its description should have a verb.

File: std/haxe/macro/Expr.hx [hunk 30]
```diff
@@ -324,24 +817,65 @@ enum TypeDefKind {
...
+	@see http://haxe.org/manual/type-system-import.html
+**/
 enum ImportMode {
+	/**
+		Represents a default import `import c`.
+	**/
 	INormal;
+
+	/**
+		Represents the alias import `import c in alias`.
```
Comment (Simn) [inline std/haxe/macro/Expr.hx]: I'm pushing for "import c as alias" to be the official thing to use.

**Comment by Simn** [comment]
Thanks!

Sorry for the nitpicking but it's quite important to be precise with terminology in this area.

File: std/haxe/macro/Expr.hx (hunk 15 above)
Comment (markknol) [inline std/haxe/macro/Expr.hx]: (LOL) Ah, I thought it was a reference to a subclass/parent type thing in case when there was inheritance. I learn a lot of this.

Comment (markknol) [inline std/haxe/macro/Expr.hx]: Can you help how to describe this?

File: std/haxe/macro/Expr.hx (hunk 16 above)
Comment (markknol) [inline std/haxe/macro/Expr.hx]: Ok same as above, I thought it was something complete different. Can you help how to describe this?

Comment (Simn) [inline std/haxe/macro/Expr.hx]: It's this thing: http://haxe.org/manual/type-system-extensions.html

The array has the type paths to the `> Iterable<T>` etc. things, the fields are just fields.

File: std/haxe/macro/Expr.hx (hunk 15 above)
Comment (Simn) [inline std/haxe/macro/Expr.hx]: "Represents parentheses around a type, e.g. the `(Int -> Void)` part in `(Int -> Void) -> String`."

File: std/haxe/macro/Expr.hx (hunk 27 above)
Comment (markknol) [inline std/haxe/macro/Expr.hx]: :smile: They all feel like _"The structure of the type kind type typed as the type kind definition structure definitions."_

File: std/haxe/macro/Expr.hx (hunk 30 above)
Comment (markknol) [inline std/haxe/macro/Expr.hx]: Is this already there (should I document like this), or is this your dream?

Comment (Simn) [inline std/haxe/macro/Expr.hx]: That works as of 3.2.0.

**Comment by markknol** [comment]
Thanks!! I already expected some feedback, so no need to apologize. The goal is to make it crystal clear.

Can you double check, especially `TypeParam`?

Also, I added a `Constant.CRegexp` example, but I'm not sure what examples should be given for the others.

**Comment by Simn** [comment]
I don't remember what I wanted to do here so I'll go ahead and merge it. Thanks a lot for your contributions!
